### DISCUSSION TITLE
Agent to use another agent as tool?

### DISCUSSION THREAD

**Discussion by yos1p** [FlowiseAI/Flowise#2181] [Ideas]
Now that we have another Chain as tool, I wonder if it's also possible to use another Agent as tool. Even better, to use another Flow as a Tool.

For example, if I want to talk about Topic A, the Agent will use the tool to "forward" the chat to Agent A. And if I want to talk about Topic B, "forward" to Agent B. Is this something possible?

**Comment by acdewinter** [comment]
This is possible (took a lot of playing around) but I have managed to make this work. Maybe I'll make a detailed post at some point but the basic workflow is:

Make 2 flows. 
Get the API (curl) details from flow 2.
Make custom tool and setup the API details from flow 2 as a JS function (you can play around yourself to get it to take the usermessage from Flow1 and give it to the tool) and return response.
Attach tool to agent on flow 1.

Obviously you need to play around with your own descriptions to optimize tool calling for Agent in flow 1.

Hope this helps :)

**Comment by toi500** [comment]
I would love to read such a post about it. Or sharing the 2 flows as templates, if a detailed article is too much work.

**Comment by sirsimonson** [comment]
My approach to this is to just have a wrapper-agent (to stream the delayed response) with a personality/role prompt which has a strict instruction to use one tool ALWAYS (which is duplicated in the tool description).

This custom tool calls two seperate chatflows in sequence. First it does classification/planning, second takes the result and feeds it to a call to another chatflow (which is an agent).

This works well but it does lack of the history beeing passed from the parent (wrapper) to the both child chatflows. 

Hope this helps and maybe someone has advice how to retrieve a chat_history inside a custom tool? 

Actually if this could be resolved (chat_history shared through flows) this could be a very powerful pattern to address an agent that does some kind of complex sequence of chatflows. In my case this makes sense because I want to classify in order to trigger a "Socratic mode", rag-mode or tool-agent retrieval to power a conversation to the user. This empowers to do complex conditional classification resulting in a complex answer that is used by my customer-facing chatflows as an "intermediate_step" tool with a human like expert output to be then paraphrased by the agent. 

-> long latency but very promising safe results.

Here is the flow which is just a tool agent equipped with one custom tool:
![intermediate_step](https://github.com/FlowiseAI/Flowise/assets/42245157/cecf5be6-a454-462c-af95-d11eddd570e4)

**Comment by amiavia** [comment]
Are there any new development on this topic? I have tried to have an agent (within an Agent Flow) triggering another chat flow (within Chatflow) which retrieves some information via RAG from a Vector DB. There is no response in the chat also it seems like the Chatflow Tool is not being called by the agent. Would appreciate any insight / idea how to get this done.
